### ISSUE TITLE
Renaming a resource while also setting ignoreChanges on the name leads to a panic

### ISSUE THREAD

**Issue by hnnsgstfssn** [pulumi/pulumi-aws#1207]
Removing bucket life-cycle rules fails with the following stack trace. The log does unfortunately not provide more useful information even with very verbose output. 
```
7604 step_executor.go:321] StepExecutor worker(23): step update on urn:pulumi:x::x.cloud::aws:s3/bucket:Bucket::storage-stage failed with an error: transport is closing
7604 step_executor.go:321] StepExecutor worker(23): step update on urn:pulumi:x::x.cloud::aws:s3/bucket:Bucket::storage-stage failed, signalling cancellation
```
```
Updating (x):
     Type                 Name           Status                  Info
     pulumi:pulumi:Stack  x.cloud-x      **failed**              1 error; 22 messages
 ~   └─ aws:s3:Bucket     storage-stage  **updating failed**     [diff: ~lifecycleRules]; 1 error
Diagnostics:
  pulumi:pulumi:Stack (x.cloud-x):
    error: update failed
    panic: fatal: An assertion has failed: Expected diff to not require deletion or replacement during Update of urn:pulumi:x::x.cloud::aws:s3/bucket:Bucket::storage-stage
    goroutine 84 [running]:
    github.com/pulumi/pulumi/sdk/v2/go/common/util/contract.failfast(...)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.12.0/go/common/util/contract/failfast.go:23
    github.com/pulumi/pulumi/sdk/v2/go/common/util/contract.Assertf(0xc0017a5600, 0x64b8c1b, 0x48, 0xc001d8b790, 0x1, 0x1)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.12.0/go/common/util/contract/assert.go:33 +0x197
    github.com/pulumi/pulumi-terraform-bridge/v2/pkg/tfbridge.(*Provider).Update(0xc0006db6c0, 0x6fca300, 0xc001c073e0, 0xc001582c80, 0xc0006db6c0, 0x569c601, 0xc001c0d380)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v2@v2.11.1-0.20201020163502-64cff1e50894/pkg/tfbridge/provider.go:958 +0x831
    github.com/pulumi/pulumi/sdk/v2/proto/go._ResourceProvider_Update_Handler.func1(0x6fca300, 0xc001c073e0, 0x61dcd60, 0xc001582c80, 0x61a1e60, 0xb322ec8, 0x6fca300, 0xc001c073e0)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.12.0/proto/go/provider.pb.go:2280 +0x89
    github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x6fca300, 0xc001c06210, 0x61dcd60, 0xc001582c80, 0xc0019269a0, 0xc0019269c0, 0x0, 0x0, 0x6f025c0, 0xc00029e070)
        /home/runner/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x2eb
    github.com/pulumi/pulumi/sdk/v2/proto/go._ResourceProvider_Update_Handler(0x6295860, 0xc0006db6c0, 0x6fca300, 0xc001c06210, 0xc001922840, 0xc000691e60, 0x6fca300, 0xc001c06210, 0xc001471300, 0x1177)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.12.0/proto/go/provider.pb.go:2282 +0x14b
    google.golang.org/grpc.(*Server).processUnaryRPC(0xc0006db500, 0x6ff0760, 0xc0010cc000, 0xc00178d400, 0xc000e25590, 0xb2e46d8, 0x0, 0x0, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1171 +0x50a
    google.golang.org/grpc.(*Server).handleStream(0xc0006db500, 0x6ff0760, 0xc0010cc000, 0xc00178d400, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1494 +0xccd
    google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000500060, 0xc0006db500, 0x6ff0760, 0xc0010cc000, 0xc00178d400)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:834 +0xa1
    created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:832 +0x204
  aws:s3:Bucket (storage-stage):
    error: transport is closing
```
Here is output from `pulumi preview --diff`.
```
Previewing update (x):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:x::x.cloud::pulumi:pulumi:Stack::x.cloud-x]
    ~ aws:s3/bucket:Bucket: (update)
        [id=x.cloud]
        [urn=urn:pulumi:x::x.cloud::aws:s3/bucket:Bucket::storage-stage]
      ~ lifecycleRules: [
          ~ [0]: {
                  ~ abortIncompleteMultipartUploadDays: 1 => 7
                  ~ expiration                        : {
                      ~ days: 30 => 7
                    }
                  ~ noncurrentVersionExpiration       : {
                      ~ days: 30 => 7
                    }
                  ~ prefix                            : "stage/" => "test2/"
                }
          ~ [1]: {
                  ~ prefix: "test2/" => "test/"
                }
          ~ [2]: {
                  ~ expiration                 : {
                      ~ days: 7 => 30
                    }
                  ~ noncurrentVersionExpiration: {
                      ~ days: 7 => 30
                    }
                  ~ prefix                     : "test/" => "prototype/"
                }
          - [3]: {
                  - abortIncompleteMultipartUploadDays: 7
                  - enabled                           : true
                  - expiration                        : {
                      - days                     : 30
                      - expiredObjectDeleteMarker: false
                    }
                  - id                                : "prod 30d retention rule"
                  - noncurrentVersionExpiration       : {
                      - days: 30
                    }
                  - prefix                            : "prod/"
                }
          - [4]: {
                  - abortIncompleteMultipartUploadDays: 7
                  - enabled                           : true
                  - expiration                        : {
                      - days                     : 30
                      - expiredObjectDeleteMarker: false
                    }
                  - id                                : "input 30d retention rule"
                  - noncurrentVersionExpiration       : {
                      - days: 30
                    }
                  - prefix                            : "prototype/"
                }
        ]
Resources:
    ~ 1 to update
    29 unchanged
Permalink:
```

**Comment by komalali** [comment]
@hnnsgstfssn I'm having trouble getting a repro for this. I tried to mirror as close as I could from what I can see in your diff, but I am being able to update as expected. If you can give us a code example for how to repro this issue that'd be greatly appreciated.

**Comment by hnnsgstfssn** [comment]
@komalali while I work on a repro, would it help to know that I've  before applying this change have renamed the bucket and notification resource? I've applied, in succession

```
-           bucket, err := s3.NewBucket(ctx, "Storage", &s3.BucketArgs{
+         bucket, err := s3.NewBucket(ctx, "storage-stage", &s3.BucketArgs{
                        LifecycleRules: s3.BucketLifecycleRuleArray{
                                &s3.BucketLifecycleRuleArgs{
                                        AbortIncompleteMultipartUploadDays: pulumi.IntPtr(1),
@@ -542,11 +555,12 @@ func main() {
                        pulumi.IgnoreChanges([]string{
                                "bucket",
                        }),
+                 pulumi.Aliases([]pulumi.Alias{{Name: pulumi.String("Storage")}}),
                )
                if err != nil {
                        return err
                }
-           _, err = s3.NewBucketNotification(ctx, "bucketNotification", &s3.BucketNotificationArgs{
+         _, err = s3.NewBucketNotification(ctx, "bucket-notification", &s3.BucketNotificationArgs{
                        Bucket: bucket.ID(),
                        LambdaFunctions: s3.BucketNotificationLambdaFunctionArray{
                                &s3.BucketNotificationLambdaFunctionArgs{
@@ -565,6 +579,7 @@ func main() {
                                },
                        },
                },
+                 pulumi.Aliases([]pulumi.Alias{{Name: pulumi.String("bucketNotification")}}),
                )
                if err != nil {
                        return err
```

and after that I've removed the alias options.

**Comment by hnnsgstfssn** [comment]
Reordering life-cycle rules in my list is not allowed either.

```
pulumi preview --diff
warning: Pulumi will not be able to print a statefile permalink using these credentials. Neither a GoogleAccessID or PrivateKey are available. Try using a GCP Service Account.
Enter your passphrase to unlock config/secrets
    (set PULUMI_CONFIG_PASSPHRASE or PULUMI_CONFIG_PASSPHRASE_FILE to remember):
Previewing update (hoxton):
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:x::x.cloud::pulumi:pulumi:Stack::x.cloud-x]
    ~ aws:s3/bucket:Bucket: (update)
        [id=leonard.hoxtonanalytics.com]
        [urn=urn:pulumi:x::x.cloud::aws:s3/bucket:Bucket::storage-stage]
      ~ lifecycleRules: [
          ~ [0]: {
                  ~ abortIncompleteMultipartUploadDays: 1 => 7
                  ~ expiration                        : {
                      ~ days: 30 => 7
                    }
                  ~ noncurrentVersionExpiration       : {
                      ~ days: 30 => 7
                    }
                  ~ prefix                            : "stage/" => "x/"
                }
          ~ [1]: {
                  ~ abortIncompleteMultipartUploadDays: 7 => 1
                  ~ expiration                        : {
                      ~ days: 7 => 30
                    }
                  ~ noncurrentVersionExpiration       : {
                      ~ days: 7 => 30
                    }
                  ~ prefix                            : "x/" => "stage/"
                }
        ]
Resources:
    ~ 1 to update
    29 unchanged
```

fails with a similar stack trace.

**Comment by hnnsgstfssn** [comment]
Creating a new project trying to import the problematic bucket fails sporadically

```
pulumi import aws:s3/bucket:Bucket x problematicbucket
Enter your passphrase to unlock config/secrets
    (set PULUMI_CONFIG_PASSPHRASE or PULUMI_CONFIG_PASSPHRASE_FILE to remember):
Previewing import (dev):
     Type                 Name   Plan
     pulumi:pulumi:Stack  x-dev  running.
 =   └─ aws:s3:Bucket     x      import.
panic: runtime error: index out of range [4] with length 4

goroutine 84 [running]:
github.com/pulumi/pulumi/sdk/v2/go/common/resource.PropertyPath.Add(0xc004b106c0, 0x3, 0x4, 0x192af20, 0xc004f1a5c0, 0x192af20, 0xc004f1a980, 0x1bd2c01, 0xc00270e7b0, 0x1)
	/Users/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:202 +0x790
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*ImportStep).Apply(0xc000aad3b0, 0x1, 0x1ce59f9, 0x23, 0xc0011cde40, 0x3)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step.go:979 +0x985
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).executeStep(0xc0001a5900, 0x2, 0x200dd60, 0xc000aad3b0, 0xc000ab9140, 0x0)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:271 +0x209
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).executeChain(0xc0001a5900, 0x2, 0xc000a91500, 0x1, 0x1)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:221 +0xef
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).worker.func1(0xc0001a5900, 0x2, 0xc000a91500, 0x1, 0x1, 0xc000799860)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:371 +0xc5
created by github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).worker
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:368 +0x2e7
```
and rerunning the import yields the same error with different indicies 
```
pulumi import aws:s3/bucket:Bucket x problematicbucket
Enter your passphrase to unlock config/secrets
    (set PULUMI_CONFIG_PASSPHRASE or PULUMI_CONFIG_PASSPHRASE_FILE to remember):
Previewing import (dev):
     Type                 Name   Plan
     pulumi:pulumi:Stack  x-dev  running.
 =   └─ aws:s3:Bucket     x      import.
panic: runtime error: index out of range [3] with length 3

goroutine 40 [running]:
github.com/pulumi/pulumi/sdk/v2/go/common/resource.PropertyPath.Add(0xc0036e2540, 0x3, 0x4, 0x192af20, 0xc000183b20, 0x192af20, 0xc000183cc0, 0x1bd2c01, 0xc004b1b920, 0x1)
	/Users/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:202 +0x790
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*ImportStep).Apply(0xc000486230, 0x1, 0x1ce59f9, 0x23, 0xc0001bde40, 0x3)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step.go:979 +0x985
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).executeStep(0xc0000d1900, 0x2, 0x200dd60, 0xc000486230, 0xc000268758, 0x0)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:271 +0x209
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).executeChain(0xc0000d1900, 0x2, 0xc000756640, 0x1, 0x1)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:221 +0xef
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).worker.func1(0xc0000d1900, 0x2, 0xc000756640, 0x1, 0x1, 0xc00099e120)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:371 +0xc5
created by github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).worker
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:368 +0x2e7
```
If I keep rerunning the import it will eventually succeed

```
pulumi import aws:s3/bucket:Bucket x problematicbucket
Enter your passphrase to unlock config/secrets
    (set PULUMI_CONFIG_PASSPHRASE or PULUMI_CONFIG_PASSPHRASE_FILE to remember):
Previewing import (dev):
     Type                 Name   Plan
     pulumi:pulumi:Stack  x-dev
 =   └─ aws:s3:Bucket     x      import

Resources:
    = 1 to import
    1 unchanged

Permalink: file:///home/x/.pulumi/stacks/dev.json
Do you want to perform this import?  [Use arrows to move, enter to select, type to filter]
  yes
> no
  details
```

**Comment by hnnsgstfssn** [comment]
Creating a new project trying to import the problematic bucket fails sporadically

```
pulumi import aws:s3/bucket:Bucket x problematicbucket
Enter your passphrase to unlock config/secrets
    (set PULUMI_CONFIG_PASSPHRASE or PULUMI_CONFIG_PASSPHRASE_FILE to remember):
Previewing import (dev):
     Type                 Name   Plan
     pulumi:pulumi:Stack  x-dev  running.
 =   └─ aws:s3:Bucket     x      import.
panic: runtime error: index out of range [4] with length 4

goroutine 84 [running]:
github.com/pulumi/pulumi/sdk/v2/go/common/resource.PropertyPath.Add(0xc004b106c0, 0x3, 0x4, 0x192af20, 0xc004f1a5c0, 0x192af20, 0xc004f1a980, 0x1bd2c01, 0xc00270e7b0, 0x1)
	/Users/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:202 +0x790
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*ImportStep).Apply(0xc000aad3b0, 0x1, 0x1ce59f9, 0x23, 0xc0011cde40, 0x3)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step.go:979 +0x985
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).executeStep(0xc0001a5900, 0x2, 0x200dd60, 0xc000aad3b0, 0xc000ab9140, 0x0)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:271 +0x209
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).executeChain(0xc0001a5900, 0x2, 0xc000a91500, 0x1, 0x1)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:221 +0xef
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).worker.func1(0xc0001a5900, 0x2, 0xc000a91500, 0x1, 0x1, 0xc000799860)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:371 +0xc5
created by github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).worker
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:368 +0x2e7
```
and rerunning the import yields the same error with different indicies 
```
pulumi import aws:s3/bucket:Bucket x problematicbucket
Enter your passphrase to unlock config/secrets
    (set PULUMI_CONFIG_PASSPHRASE or PULUMI_CONFIG_PASSPHRASE_FILE to remember):
Previewing import (dev):
     Type                 Name   Plan
     pulumi:pulumi:Stack  x-dev  running.
 =   └─ aws:s3:Bucket     x      import.
panic: runtime error: index out of range [3] with length 3

goroutine 40 [running]:
github.com/pulumi/pulumi/sdk/v2/go/common/resource.PropertyPath.Add(0xc0036e2540, 0x3, 0x4, 0x192af20, 0xc000183b20, 0x192af20, 0xc000183cc0, 0x1bd2c01, 0xc004b1b920, 0x1)
	/Users/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:202 +0x790
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*ImportStep).Apply(0xc000486230, 0x1, 0x1ce59f9, 0x23, 0xc0001bde40, 0x3)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step.go:979 +0x985
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).executeStep(0xc0000d1900, 0x2, 0x200dd60, 0xc000486230, 0xc000268758, 0x0)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:271 +0x209
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).executeChain(0xc0000d1900, 0x2, 0xc000756640, 0x1, 0x1)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:221 +0xef
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).worker.func1(0xc0000d1900, 0x2, 0xc000756640, 0x1, 0x1, 0xc00099e120)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:371 +0xc5
created by github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepExecutor).worker
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_executor.go:368 +0x2e7
```
If I keep rerunning the import it will eventually succeed

```
pulumi import aws:s3/bucket:Bucket x problematicbucket
Enter your passphrase to unlock config/secrets
    (set PULUMI_CONFIG_PASSPHRASE or PULUMI_CONFIG_PASSPHRASE_FILE to remember):
Previewing import (dev):
     Type                 Name   Plan
     pulumi:pulumi:Stack  x-dev
 =   └─ aws:s3:Bucket     x      import

Resources:
    = 1 to import
    1 unchanged

Permalink: file:///home/x/.pulumi/stacks/dev.json
Do you want to perform this import?  [Use arrows to move, enter to select, type to filter]
  yes
> no
  details
```

**Comment by joeduffy** [comment]
I've tried but haven't been able to reproduce this.

I suspect there is a good chance `lifecycleRules` is a red herring. There is nothing about `lifecycleRules` that would trigger a replacement. In fact, the only ForceNew properties for S3 Buckets are `bucket`, `bucketPrefix`, and `objectLockConfiguration.objectLockEnabled`.

I noticed this in the code above

```
                        pulumi.IgnoreChanges([]string{
                                "bucket",
                        }),
```

and the fact that there was a rename operation here makes me very suspicious.

Is it possible something about the rename operation is causing Pulumi to try changing the `bucket` property during the update (unexpectedly and in a way not shown in the diff), which would presumably trigger this panic?

As an aside, the import errors appear to be an entirely different bug, which we have seen elsewhere: https://github.com/pulumi/pulumi-aws/issues/1245.

**Comment by hnnsgstfssn** [comment]
@joeduffy Thanks for taking another look at this. I was also unable to reproduce it and stopped trying as I kept running into the other issue mentioned above. I noticed that a [fix for that](https://github.com/pulumi/pulumi-terraform-bridge/pull/304) was merged just the other day and I'm waiting for a release that contains it.

Regarding `pulumi.IgnoreChanges([]string{"bucket",}),`, when it is present I consistently get the same error originally reported, trying to modify the retention rules. However when removing it and instead setting the bucket attribute the error disappears and the changes apply successfully!

**Comment by komalali** [comment]
@hnnsgstfssn the fix you mentioned has been released in pulumi-aws 3.19.2. Let me know if that resolves the issue!

**Comment by hnnsgstfssn** [comment]
@komalali ah, thanks. I've just tried importing the resource and I'm no longer seeing the import errors. The bucket imports successfully.

**Comment by hnnsgstfssn** [comment]
@joeduffy I recon you're spot on with the root cause being unrelated to this resource. I've just tried to modify, in the same project, a lambda and change the environment variables. The lambda has the `pulumi.IgnoreChanges([]string{"name"})` option applied and I get

```
panic: fatal: An assertion has failed: Expected diff to not require deletion or replacement during Update of urn:pulumi:x::x.cloud::aws:lambda/function:Function::upload-stage
```

Removing the ignore and instead defining the attribute then correctly diffs and applying the change succeeds.

This lambda resource has not previously been renamed.

**Comment by stack72** [comment]
I am going to close this out as it seems this was fixed in 3.19.2

if this is still a problem, please let us know and we can re-investigate this
